addons/l10n_vn_edi/models/account_move.py

```diff
@@ -0,0 +1,719 @@
+        sinvoices.l10n_vn_edi_sinvoice_xml_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_pdf_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_file_id.unlink()
+        return super().button_draft()
+
+    def _l10n_vn_edi_generate_invoice_json(self):
+        """ Return the dict of data that will be sent to the api in order to create the invoice. """
+        # We leave the summarized information computation to SInvoice.
+        self.ensure_one()
+        # This MUST match chronologically with the sequence they generate on their system, which is why it is set to now.
+        self.l10n_vn_edi_issue_date = fields.Datetime.now()
+        return {
+            'generalInvoiceInfo': self._l10n_vn_edi_get_general_invoice_information(),
+            'buyerInfo': self._l10n_vn_edi_get_buyer_information(),
+            'sellerInfo': self._l10n_vn_edi_get_seller_information(),
+            'payments': self._l10n_vn_edi_get_payment_information(),
+            'itemInfo': self._l10n_vn_edi_get_item_information(),
+            'taxBreakdowns': self._l10n_vn_edi_get_tax_breakdowns(),
+        }
+
+    def _l10n_vn_edi_get_general_invoice_information(self):
+        """ General invoice information, such as the model number, invoice symbol, type, date of issues, ... """
+        self.ensure_one()
+        invoice_data = {
+            'transactionUuid': str(uuid.uuid4()),
+            'invoiceType': self.l10n_vn_edi_invoice_symbol.invoice_template_id.template_invoice_type,
+            'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+            'invoiceSeries': self.l10n_vn_edi_invoice_symbol.name,
+            # This timestamp is important as it is used to check the chronological order of Invoice Numbers.
+            # Since this xml is generated upon posting, just like the invoice number, using now() should keep that order
+            # correct in most case.
+            'invoiceIssuedDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+            'currencyCode': self.currency_id.name,
+            'adjustmentType': '1',  # 1 for original invoice, which is the case during first issuance.
+            'paymentStatus': self.payment_state in {'in_payment', 'paid'},
+            'cusGetInvoiceRight': True,  # Set to true, allowing the customer to see the invoice.
+            'validation': 1,  # Set to 1, SInvoice will validate tax information while processing the invoice.
+        }
+
+        # When invoicing in a foreign currency, we need to provide the rate, or it will default to 1.
+        if self.currency_id.name != 'VND':
+            invoice_data['exchangeRate'] = self.env['res.currency']._get_conversion_rate(
+                from_currency=self.company_currency_id,
+                to_currency=self.currency_id,
+                company=self.company_id,
+                date=self.invoice_date or self.date or fields.Date.context_today(self),
```

Comment:
At this point, the move is saved so `self.date` is always there (required field)